### PR TITLE
Add functionality to read compressed gzip files.

### PR DESCRIPTION
**Pull request by gassmoeller**
As the changelog says: 
The `read_and_distribute_file_content` utility function can now also read
binary files compressed with gzip, which allows for example the StructuredData
class (and therefore all ascii data plugins) to read gzip compressed data.
Compressed files are detected based on their file ending '.gz'. I.e. if you want
to read a compressed file, make sure it's name ends in '.gz', and if you want to
read a plain text file, make sure its name does not end in '.gz'. Compressed
files can save around 75% to 80% of disk space for typical ascii data files.

### CODE REVIEW

File: source/utilities.cc [hunk 1]
```diff
@@ -1363,21 +1371,26 @@ namespace aspect
...
-
               // Read data from disk
               std::stringstream datastream;
-              filestream >> datastream.rdbuf();
 
-              if (!filestream.eof())
+              try
+                {
+                  boost::iostreams::filtering_istreambuf in;
+                  if (filename.find(".gz") == filename.size()-3)
```
Comment (tjhei) [inline source/utilities.cc:1380]: What if my file is called ``test.gzinga.txt.gz``?

File: doc/modules/changes/20220520_gassmoeller [hunk 2]
```diff
@@ -0,0 +1,9 @@
+New: The `read_and_distribute_file_content` utility function can now also read
+binary files compressed with gzip, which allows for example the StructuredData
+class (and therefore all ascii data plugins) to read gzip compressed data.
+Compressed files are detected based on their file ending '.gz'. I.e. if you want
+to read a compressed file, make sure it's name ends in '.gz', and if you want to
```
Comment (bangerth) [inline doc/modules/changes/20220520_gassmoeller:5]:
```suggestion
to read a compressed file, make sure its name ends in '.gz', and if you want to
```

File: source/utilities.cc [hunk 3]
```diff
@@ -1351,7 +1355,11 @@ namespace aspect
             }
           else
             {
-              std::ifstream filestream(filename.c_str());
+              std::ifstream filestream;
+              if (filename.find(".gz") == filename.size()-3)
```
Comment (bangerth) [inline source/utilities.cc:1359]: Maybe easier to read, at least for those of us with magic regex skills :-)
```suggestion
              if (std::regex_match (filename, std::regex("\\.gz$"))
```

File: source/utilities.cc (hunk 1 above)
Comment (bangerth) [inline source/utilities.cc:1380]: same here. or maybe introduce a local variable that stores whether you're dealing with a compressed file.

**Review by bangerth — COMMENTED** [review]
Very nice!

**Review by bangerth — APPROVED** [review]

**Review by tjhei — APPROVED** [review]

**Pull request merged**